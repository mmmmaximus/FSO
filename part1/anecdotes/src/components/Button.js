import React from 'react'

const Button = ({text, onClick}) => <div data-testid='Button'><button onClick={onClick}>{text}</button></div>

export default Button
