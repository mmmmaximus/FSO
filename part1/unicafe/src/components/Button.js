import React from 'react'

const Button = ({text, onClick}) => <button onClick={onClick} data-testid='Button'>{text}</button>

export default Button
