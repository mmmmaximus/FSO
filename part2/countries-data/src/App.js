import React, {useState, useEffect} from 'react';
import axios from 'axios';
import Filter from './components/Filter'
import Countries from './components/Countries'

const App = () => {
  const [filter, setFilter] = useState('')
  const [countries, setCountries] = useState([])
  
  const changeFilter = (event) => {
    setFilter(event.target.value)
  }

  useEffect(() => {
    axios
      .get('https://restcountries.eu/rest/v2/all')
      .then(response => {
        setCountries(response.data)
      })
  }, [])
  
  return (
    <div>
      <Filter func={changeFilter}/>
      <Countries 
        countries={
          countries.filter(
            country => country.name.toLowerCase().includes(filter.toLowerCase())
          )
        }
        filter={filter}
      />
    </div>
  )
}

export default App;
