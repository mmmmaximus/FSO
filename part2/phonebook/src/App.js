import React, { useState, useEffect } from 'react'
import Input from './components/Input'
import PersonForm from './components/PersonForm'
import Persons from './components/Persons'
import ErrorMessage from './components/ErrorMessage'
import phonebookService from './services/phonebook'

const App = () => {
  const [persons, setPersons] = useState([])
  const [newFilter, setNewFilter] = useState('')
  const [newName, setNewName] = useState('')
  const [newNumber, setNewNumber] = useState('')
  const [errorMessage, setErrorMessage] = useState('')
  const [messageType, setMessageType] = useState('')
  const changeFilter = (event) => {
    setNewFilter(event.target.value)
  }
  const changeNumber = (event) => {
    setNewNumber(event.target.value)
  }
  const changeName = (event) => {
    setNewName(event.target.value)
  }
  const clearInputs = () => {
    setNewName('')
    setNewNumber('')
    document.getElementById('name').value = ''
    document.getElementById('number').value = ''
  }

  useEffect(() => {
    phonebookService
      .index()
      .then(persons => setPersons(persons))
  }, [])

  const addContact = (event) => {
    event.preventDefault()
    if (newName === '' || newNumber === '') {
      alert(`Name and Number must be filled in`)
    } else if (persons.map(person => person.name.toLowerCase()).includes(newName.toLowerCase())) {
      if (persons.map(person => person.number).includes(newNumber)) {
        alert(`${newName} is already added to phonebook`)
      } else {
        if (window.confirm(`${newName} is already added to phonebook, replace old number?`)) {
          const changedPerson = persons.find(person => person.name === newName)
          phonebookService
            .update(changedPerson.id, { name: newName, number: newNumber })
            .then(person => {
              setPersons(persons.filter(person => person.name !== newName).concat(person))
              setMessageType('notification')
              setErrorMessage(`${person.name} has been updated`)
              setTimeout(() => {
                setErrorMessage('')
              }, 1000)
            }).catch(err => {
              setMessageType('error')
              setErrorMessage(`${newName} has already been deleted`)
              setTimeout(() => {
                setErrorMessage('')
              }, 1000)
            })
          clearInputs()
        }
      }
    } else {
      phonebookService
        .create({ name: newName, number: newNumber })
        .then(person => {
          setPersons(persons.concat(person))
          setMessageType('notification')
          setErrorMessage(`${person.name} has been added`)
          setTimeout(() => {
            setErrorMessage('')
          }, 1000)
        })
      clearInputs()
    }
  }

  const deletePerson = (id, name) => {
    if (window.confirm(`Delete ${name}?`)) {
      phonebookService
        .del(id)
        .then(phonebookService
          .index()
          .then(persons => {
            setPersons(persons.filter(person => person.id !== id))
            setMessageType('notification')
            setErrorMessage(`${name} has been deleted`)
            setTimeout(() => {
              setErrorMessage('')
            }, 1000)
          })
        )
    }
  }

  return (
    <div>
      <h2>Phonebook</h2>
      <ErrorMessage message={errorMessage} type={messageType}/>
      <Input text='filter shown with ' func={changeFilter} />
      <h2>add a new</h2>
      <PersonForm changeName={changeName} changeNumber={changeNumber} addContact={addContact}/>
      <h2>Numbers</h2>
      <Persons deletePerson={deletePerson} persons={
        persons.filter(
          person => person.name.toLowerCase().includes(newFilter.toLowerCase())
        )
      }/>
      ...
    </div>
  )
}

export default App
