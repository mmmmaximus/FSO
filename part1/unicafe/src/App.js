import React, { useState } from 'react'
import Header from './components/Header'
import Button from './components/Button'
import Statistics from './components/Statistics'

const App = () => {
  const [good, setGood] = useState(0)
  const [neutral, setNeutral] = useState(0)
  const [bad, setBad] = useState(0)

  return (
    <div data-testid='App'>
      <Header header='give feedback'/>
      <Button text='good' onClick={() => setGood(good +1)}/>
      <Button text='neutral' onClick={() => setNeutral(neutral +1)}/>
      <Button text='bad' onClick={() => setBad(bad +1)}/>
      <Header header='statistics'/>
      <Statistics good={good} neutral={neutral} bad={bad}/>
    </div>
  )
}

export default App
