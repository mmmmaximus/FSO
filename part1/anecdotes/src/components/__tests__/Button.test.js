import { fireEvent, render, screen } from '@testing-library/react'
import Button from './../Button'
import App from './../../App'
import renderer from 'react-test-renderer'
import { useState } from 'react'

test('Button component renders on screen', () => {
  const mockfn = jest.fn()
  render(<Button text='test text' onClick={mockfn}/>)
  const ButtonElement = screen.getByText('test text')
  expect(ButtonElement).toBeInTheDocument
  expect(ButtonElement.textContent).toBe('test text')

  // calls func when clicked
  fireEvent.click(ButtonElement)
  expect(mockfn.mock.calls).toHaveLength(1)
})

test('vote button works', () => {
  render(<App />)
  const AppElement = screen.getByTestId('App')
  const voteButton = screen.getByText('vote')
  expect(AppElement.textContent).toContain('has 0 votes')

  fireEvent.click(voteButton)
  expect(AppElement.textContent).toContain('has 1 votes')
  expect(AppElement.textContent).not.toContain('has 0 votes')

  // check if next anecdote button works, using enzyme?
})

it('matches snapshot', () => {
  const mockfn = jest.fn()
  const tree = renderer.create(<Button text='test text' onClick={mockfn} />).toJSON()
  expect(tree).toMatchSnapshot()
})
