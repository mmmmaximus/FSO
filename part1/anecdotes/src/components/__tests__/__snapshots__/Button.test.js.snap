// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  data-testid="Button"
>
  <button
    onClick={[MockFunction]}
  >
    test text
  </button>
</div>
`;
